### PR TITLE
#17134998 Two factor authentication

### PR DESCRIPTION
**Pull request by gildniy**
#### What does this PR do?
Creates a 2 factor authentication for the user.

#### Description of Task to be completed?
- [x] add the field for the `twoFAType` on the users table
- [x] add the field for the `twoFASecret` on the users table
- [x] add the field for the `twoFADataURL` on the users table
- [x] add the functionality to optionally setup the 2 fa for the logged in user and this by using authenticator application (like google authenticator or Authy) or SMS text message
- [x] verify if the user can use the 2fa before letting him activate this feature to his account
- [x] add the option for the user to know if he has activated 2fa or not
- [x] add a way for the user to disable 2fa for his account

#### How should this be manually tested?
- [x] clone the repo and run `npm install` to install the dependencies
- [x] make database migrate `npm run migrate` and  database seed `npm run db:seed`
- [x] start the app by `npm start`
* Using postman: 
- [x] login one of the users (eg. `{
    "email": "requester@user.com",
    "password": "12345678"
  }`)
* With the logged in user token in the Authentication header in postman:
- [x] to set the 2FA do a `PATCH request` to `{{url}}/api/v1/2fa/totp/setup` with body:`{
	"twoFAType": "authenticator_app_temp"
}` to be able to use the 2FA with 'Authenticator app' or `{
	"twoFAType": "sms_text_temp"
}` to be able to use the 2FA with 'SMS Text'
- [x] to check the 2FA do a `GET request` to `{{url}}/api/v1/2fa/totp/setup`
- [x] to send the 60 second valid 2FA token to the user using SMS do a `POST request` to `{{url}}/api/v1/2fa/totp/send-token-text` with in body the logged in user phoneNumber and the secret created at the time of setup of 2FA (on the first step of this section) (eg. `{
	"secret": "KBYWC6T2KB4DQPB4MRSHC4CIN5BUOSKB",
	"phoneNumber": "+250788888888"
}`) (N.B: the SMS is only sent in a PRODUCTION ENVIRONMENT)
- [x] to get the 60 seconds valid 2FA token to the user using ATHENTICATOR APP, copy the `twoFADataURL` value (gotten from the first step of this section) and view it using online QR code generator. Using a smart phone with `Google Authenticator` installed scan the QR Code to get the corresponding token
- [x] to verify the token do a `POST request` to `{{url}}/api/v1/2fa/totp/verify`
- [x] to disable the 2FA for a logged in user do a `PATCH request` to `{{url}}/api/v1/2fa/totp/disable`

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#17134998](https://www.pivotaltracker.com/story/show/171374998)

#### Screenshots (if appropriate)
1. Setup 2FA using SMS Text method
image

2. SMS message sent to the user with a 6 digit 60 seconds valid token
image

image

3. Setup 2FA using Authenticator app method (a dataURL string is create to make QR code image)
image

image

4. Validate the user if he is the owner of the account using a received 2fa token
image

5. Check if the user activated 2fa and the activated type
image

6. Disable 2fa from the user account
image

#### Questions:
N/A

### CODE REVIEW

File: src/migrations/20191205132450-user.js [hunk 1]
```diff
@@ -1,8 +1,35 @@
 module.exports = {
```
Comment (RaymondMwaura) [inline src/migrations/20191205132450-user.js]: I think you should rename this migration to something more descriptive like: '20191205132450-add-2FA-to-users.js'

File: src/controllers/2fa.controller.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  const tokenMethod = type.includes('sms_text') ? { phoneNumber } : { twoFADataURL: dataURL };
+
+  const data = { twoFAType: type, twoFASecret: secret, ...tokenMethod };
+
+  if (!isTokenValid) {
+    return handleSuccess(400, 'Invalid TOTP token', res, { ...data, isTokenValid });
+  }
+
+  if (type.includes('_temp')) {
+    await db.user.update({ twoFAType: type.split('_temp')[0] }, { where: { email: res.locals.user.email } });
```
Comment (RaymondMwaura) [inline src/controllers/2fa.controller.js]: I think you should move this to you 2fa service.

File: src/controllers/2fa.controller.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  return handleSuccess(200, 'Valid TOTP token', res, { ...data,
+    isTokenValid,
+  });
+};
+
+const totpSendTokenText = async (req, res) => {
+  const { secret, phoneNumber } = req.body;
+  const { token } = generate(secret);
+
+  await messenger(phoneNumber, `Your 6 Digit 60 seconds expiration PassCode is: ${token}`);
```
Comment (RaymondMwaura) [inline src/controllers/2fa.controller.js]: I think you can create a variable from this string: ``Your 6 Digit 60 seconds expiration PassCode is: ${token}``, then call it on lines 69 & 77.

**Review by RaymondMwaura — CHANGES_REQUESTED** [review]
Great work Gildas. I have left some comments on some files on changes I think you should make. Also, add validation schema for the these two routes:
- {{{url}}/api/v1/2fa/totp/send-token-text
- {{url}}/api/v1/2fa/totp/verify

File: src/migrations/20191205132450-user.js (hunk 1 above)
Comment (gildniy) [reply src/migrations/20191205132450-user.js]: You're right, I have to put them in their own file

File: src/controllers/2fa.controller.js (hunk 2 above)
Comment (gildniy) [reply src/controllers/2fa.controller.js]: Thanks, let me move this.

**Comment by gildniy** [comment]
> Great work Gildas. I have left some comments on some files on changes I think you should make. Also, add validation schema for the these two routes:
> 
> * {{{url}}/api/v1/2fa/totp/send-token-text
> * {{url}}/api/v1/2fa/totp/verify

Here I implement the `validation` to `{{url}}/api/v1/2fa/totp/verify` only, as the other route `{{{url}}/api/v1/2fa/totp/send-token-text` the user doesn't directly enter the `phoneNumber and secret` in a form, they are provided from the backend token.

**Review by RaymondMwaura — APPROVED** [review]

**Review by RaymondMwaura — DISMISSED** [review]
Great job Gildas! All the endpoints work and all the feedback has been implemented.

**Review by nshaphan — DISMISSED** [review]
Great work @gildniy Tested, The feature works fine

**Review by b0nbon1 — DISMISSED** [review]
Hey @gildniy Great, I think everything is working okay. May be you would remove the files in the `public/temp` the PR will be good to go

**Comment by b0nbon1** [comment]
Your build is failing @gildniy

**Comment by gildniy** [comment]
> Hey @gildniy Great, I think everything is working okay. May be you would remove the files in the `public/temp` the PR will be good to go

Done

**Comment by gildniy** [comment]
> Your build is failing @gildniy

Fixed.

**Review by b0nbon1 — DISMISSED** [review]

**Review by RaymondMwaura — DISMISSED** [review]
Good job.

**Review by nshaphan — DISMISSED** [review]

**Review by RaymondMwaura — APPROVED** [review]
LGTM

**Review by nshaphan — APPROVED** [review]

**Review by b0nbon1 — APPROVED** [review]

**Pull request merged**